core/metrics-core-service/src/main/java/org/hawkular/metrics/core/service/MetricsServiceImpl.java

```diff
@@ -655,7 +656,16 @@ public void setDefaultTTL(int defaultTTL) {
                         .doOnNext(i -> insertedDataPointEvents.onNext(metric)))
                 .doOnNext(meter::mark);
 
-        Observable<Integer> indexUpdates = dataAccess.updateMetricsIndex(metrics)
+        Observable<Metric<T>> updateIdxMetrics = metrics.filter(metric -> {
+            if (metricsIdxCache.getIfPresent(metric.getMetricId()) == null) {
+                metricsIdxCache.put(metric.getMetricId(), true);
```

Comment:
There are side-effects on the cache just by reading from it. I am not really worried here about the implications on the cache itself. Also the cache is thread safe so regardless of the threading model here the cache operations will work.